### PR TITLE
[Rebase&FF] Apply memory protections

### PR DESCRIPTION
**Pull request by apop5**
## Description

During removal of Rp on Freed memory, some corner cases were also removed.

Apply the corner cases to allow for clean dxe paging audit.

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested
Q35 DxePagingAuditTestApp was reporting missing memory protection regions.
After fixes, DxePagingAuditTestApp reports passing.

## Integration Instructions
No integration necessary.

### CODE REVIEW

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 1]
```diff
@@ -764,15 +764,23 @@ GetPermissionAttributeForMemoryType (
...
-    return EFI_MEMORY_XP;
   }
 
-  return 0;
+  if (GetDxeMemoryTypeSettingFromBitfield (MemoryType, gDxeMps.NxProtectionPolicy)) {
+    Attributes |= EFI_MEMORY_XP;
+  }
+
+  if (MemoryType == EfiConventionalMemory) {
+    Attributes |= EFI_MEMORY_RP;
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: This part we shouldn't have, right? This will add RP on free memory which we dropped

File: UefiCpuPkg/CpuDxe/CpuDxe.c [hunk 2]
```diff
@@ -1048,6 +1048,15 @@ InitializeCpu (
   //
   RefreshGcdMemoryAttributes ();
 
+  // MU_CHANGE START: Install blank protocol to signal the end of the GCD sync
+  gBS->InstallMultipleProtocolInterfaces (
+         &ImageHandle,
+         &gEdkiiGcdSyncCompleteProtocolGuid,
```
Comment (os-d) [inline UefiCpuPkg/CpuDxe/CpuDxe.c]: I thought this wasn't required without RP on free memory? My recollection was that this was added for that case. I see we still have the core doing it, but I think we need to revert that commit. Checking.

Comment (os-d) [reply UefiCpuPkg/CpuDxe/CpuDxe.c]: Yeah, it was: https://github.com/microsoft/mu_basecore/commit/144e3fe63244e0dfb14a6442dd969dd2757e844e. I think we need to strip this out, it should have been removed when the RP on free memory commit was dropped. I can help if needed.

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 3]
```diff
@@ -499,69 +497,69 @@ ProtectUefiImage (
...
+   //
+   SetUefiImageProtectionAttributes (ImageRecord);
+
+   //
+   // Record the image record in the list so we can undo the protections later
+   //
+   InsertTailList (&mProtectedImageRecordList, &ImageRecord->Link);
+
+ Finish:
+   return;
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: All of this whitespace looks uneeded

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 4]
```diff
@@ -584,7 +582,6 @@ ProtectUefiImage (
   }
 
   ImageRecord = AllocateZeroPool (sizeof (*ImageRecord));
-
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c:588]: not needed

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 5]
```diff
@@ -763,15 +763,19 @@ GetPermissionAttributeForMemoryType (
...
-  } else if (GetDxeMemoryTypeSettingFromBitfield (MemoryType, gDxeMps.NxProtectionPolicy)) {
-    return EFI_MEMORY_XP;
   }
 
-  return 0;
+  if (GetDxeMemoryTypeSettingFromBitfield (MemoryType, gDxeMps.NxProtectionPolicy)) {
+    Attributes |= EFI_MEMORY_XP;
+  }
+
+  return Attributes;
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: What's the delta here? Not sure why this change is needed.

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 6]
```diff
@@ -1603,7 +1607,7 @@ ApplyMemoryProtectionPolicy (
   //
   // MU_CHANGE START Update to use memory protection settings HOB
   // if (PcdGet64 (PcdDxeNxMemoryProtectionPolicy) == 0) {
-  if (!gDxeMps.NxProtectionPolicy.Data) {
+  if (gDxeMps.NxProtectionPolicy.Data == 0) {
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: Same?

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 7]
```diff
@@ -1132,8 +1135,10 @@ MemoryProtectionCpuArchProtocolNotify (
   //
   HeapGuardCpuArchProtocolNotify ();
 
-  /*
-  if (mImageProtectionPolicy == 0) {
+  // MU_CHANGE - Start - Consume Memory Protection Hob
+  // if (mImageProtectionPolicy == 0) {
+  if (gDxeMps.ImageProtectionPolicy.Data) {
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: Should this be == 0?

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c (hunk 5 above)
Comment (apop5) [reply MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: droppd

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 8]
```diff
@@ -1116,10 +1118,10 @@ MemoryProtectionCpuArchProtocolNotify (
   //
   // Apply the memory protection policy on non-BScode/RTcode regions.
   //
-  if (PcdGet64 (PcdDxeNxMemoryProtectionPolicy) != 0) {
+  //  if (PcdGet64 (PcdDxeNxMemoryProtectionPolicy) != 0) {
+  if (!gDxeMps.NxProtectionPolicy.Data) {
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: Shouldn't this be if != 0?

**Comment by makubacki** [comment]
@apop5, can you please notify me when you have pushed the final set of expected changes to the PR?

**Comment by apop5** [comment]
@makubacki 

this PR is ready, and there is another one in mu_plus to get paging audit working on arm 5 level page tables.

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 9]
```diff
@@ -1120,10 +1118,10 @@ MemoryProtectionCpuArchProtocolNotify (
   //
   // Apply the memory protection policy on non-BScode/RTcode regions.
   //
-  if (PcdGet64 (PcdDxeNxMemoryProtectionPolicy) != 0) {
+  //  if (PcdGet64 (PcdDxeNxMemoryProtectionPolicy) != 0) {
+  if (  gDxeMps.NxProtectionPolicy.Data != 0) {
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]:
```suggestion
  if ( gDxeMps.NxProtectionPolicy.Data != 0) {
```

**Review by os-d — APPROVED** [review]

File: MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c [hunk 10]
```diff
@@ -1120,10 +1118,10 @@ MemoryProtectionCpuArchProtocolNotify (
   //
   // Apply the memory protection policy on non-BScode/RTcode regions.
   //
-  if (PcdGet64 (PcdDxeNxMemoryProtectionPolicy) != 0) {
+  //  if (PcdGet64 (PcdDxeNxMemoryProtectionPolicy) != 0) {
+  if ( gDxeMps.NxProtectionPolicy.Data != 0) {
```
Comment (makubacki) [inline MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c]: It looks like these lines might need their own MU_CHANGE tags now.

File: ArmPkg/Drivers/CpuDxe/CpuDxe.c
Comment (makubacki) [inline ArmPkg/Drivers/CpuDxe/CpuDxe.c:1]: Can you please include a brief explanation for the revert in each of the revert commit messages (for future context in history)?

**Review by makubacki — APPROVED** [review]

File: ArmPkg/Drivers/CpuDxe/CpuDxe.c
Comment (apop5) [reply ArmPkg/Drivers/CpuDxe/CpuDxe.c:1]: Both Commits have been updated with the same information.

**Review by os-d — APPROVED** [review]

**Review by os-d — APPROVED** [review]
